### PR TITLE
Fix deletion of constrained variables for CachingOptimizer

### PR DESCRIPTION
**Pull request by blegat**
To delete constrained variables, you need to delete the whole vector of variables at once if the set dimension cannot change like `MOI.SecondOrderCone`. For this reason, it is important for `CachingOptimizer` to implement the deletion of vector of variables instead of relying on the fallback functions which deletes variable one by one otherwise you get the error:
```
delete_soc_variables: Error During Test at /home/blegat/.julia/dev/MathOptInterface/src/Test/config.jl:46
  Got exception outside of a @test
  MathOptInterface.DeleteNotAllowed{MathOptInterface.VariableIndex}: Deleting the index MathOptInterface.VariableIndex(1) cannot be performed: Cannot delete variable as it is constrained with other variables in a `MOI.VectorOfVariables`. You may want to use a `CachingOptimizer` in `AUTOMATIC` mode or you may need to call `reset_optimizer` before doing this operation if the `CachingOptimizer` is in `MANUAL` mode.
  Stacktrace:
   [1] throw_delete_variable_in_vov(::MathOptInterface.VariableIndex) at /home/blegat/.julia/dev/MathOptInterface/src/Utilities/model.jl:137
   [2] _vector_of_variables_with(::Array{Tuple{MathOptInterface.ConstraintIndex{MathOptInterface.VectorOfVariables,MathOptInterface.SecondOrderCone},MathOptInterface.VectorOfVariables,MathOptInterface.SecondOrderCone},1}, ::MathOptInterface.VariableIndex) at /home/blegat/.julia/dev/MathOptInterface/src/Utilities/model.jl:148
   [3] #152 at /home/blegat/.julia/dev/MathOptInterface/src/Utilities/model.jl:175 [inlined]
   [4] broadcastvcat(::getfield(MathOptInterface.Utilities, Symbol("##152#154")){MathOptInterface.VariableIndex}, ::MathOptInterface.Utilities.ModelVectorConstraints{Float64,MathOptInterface.VectorOfVariables}) at /home/blegat/.julia/dev/MathOptInterface/src/Utilities/model.jl:964
   [5] broadcastvcat(::Function, ::MathOptInterface.Utilities.Model{Float64}) at /home/blegat/.julia/dev/MathOptInterface/src/Utilities/model.jl:935
   [6] delete(::MathOptInterface.Utilities.Model{Float64}, ::MathOptInterface.VariableIndex) at /home/blegat/.julia/dev/MathOptInterface/src/Utilities/model.jl:175
   [7] delete(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.Model{Float64}}, ::MathOptInterface.VariableIndex) at /home/blegat/.julia/dev/MathOptInterface/src/Utilities/cachingoptimizer.jl:375
   [8] delete at /home/blegat/.julia/dev/MathOptInterface/src/indextypes.jl:126 [inlined]
   [9] delete_soc_variables(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.Model{Float64}}, ::MathOptInterface.Test.TestConfig{Float64}) at /home/blegat/.julia/dev/MathOptInterface/src/Test/UnitTests/variables.jl:146
   [10] macro expansion at /home/blegat/.julia/dev/MathOptInterface/src/Test/config.jl:47 [inlined]
   [11] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Test/src/Test.jl:1113 [inlined]
   [12] unittest(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.Model{Float64}}, ::MathOptInterface.Test.TestConfig{Float64}, ::Array{String,1}) at /home/blegat/.julia/dev/MathOptInterface/src/Test/config.jl:47
   [13] unittest(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.Model{Float64}}, ::MathOptInterface.Test.TestConfig{Float64}) at /home/blegat/.julia/dev/MathOptInterface/src/Test/config.jl:42
   [14] top-level scope at /home/blegat/.julia/dev/MathOptInterface/test/Utilities/cachingoptimizer.jl:419
   [15] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Test/src/Test.jl:1113
   [16] top-level scope at /home/blegat/.julia/dev/MathOptInterface/test/Utilities/cachingoptimizer.jl:419
   [17] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.2/Test/src/Test.jl:1186
   [18] top-level scope at /home/blegat/.julia/dev/MathOptInterface/test/Utilities/cachingoptimizer.jl:393
```

### CODE REVIEW

File: src/Test/UnitTests/variables.jl [hunk 1]
```diff
@@ -88,6 +88,86 @@ function delete_variables(model::MOI.ModelLike, config::TestConfig)
...
+    MOI.delete(model, v)
+    @test MOI.get(model, MOI.NumberOfVariables()) == 0
+    @test !MOI.is_valid(model, v[1])
+    @test_throws MOI.InvalidIndex(v[1]) MOI.delete(model, v[1])
+    @test !MOI.is_valid(model, v[2])
+    @test_throws MOI.InvalidIndex(v[2]) MOI.delete(model, v[2])
+    @test !MOI.is_valid(model, cv)
+    v, cv = MOI.add_constrained_variables(model, MOI.SecondOrderCone(3))
+    @test MOI.get(model, MOI.NumberOfVariables()) == 3
+    @test_throws MOI.DeleteNotAllowed MOI.delete(model, v[1])
```
Comment (odow) [inline src/Test/UnitTests/variables.jl]: Hmmm.... I hadn't thought of this case. I guess it's reasonable.

Comment (blegat) [reply src/Test/UnitTests/variables.jl]: Passing existing tests with bridged variables put a lot of stress on the deletion API so as part of adding the variable bridges, we clarified what should be done in these corner cases. The relevant PR is https://github.com/JuliaOpt/MathOptInterface.jl/pull/793/
